Implement asymmetric encryption, signing, and key generation/import/export

Xanewok opened the pull request:
This is just a tracking PR with remaining changes I'd love to merge that are needed for [Sequoia](https://gitlab.com/sequoia-pgp/sequoia).

Xanewok committed:
Handle more CNG errors

Unfortunately `NTE_*` errors also crop up when using just `BCrypt*`
functions (rather than `NCrypt*` ones)

Xanewok committed:
Make HashAlgorithmId::to_str public

Xanewok committed:
Allow creation of an empty WindowsString

This is only to be used internally, for storage of a wide string that
can be referenced by C-compatible args for the FFI.

Xanewok committed:
WIP: Implement importing keys just from blob payload itself

Xanewok committed:
Implement RSA encryption/decryption

Xanewok committed:
Implement simple secret agreement functionality

Xanewok committed:
Implement hash signature generation/verification

Xanewok committed:
Add asymmetric algorithms, key generation and import/export

Xanewok committed:
Move KeyHandle to helpers module

This is done because it will be shared between both asymmetric and
symmetric keys.

Xanewok commented:
This should be mostly good to go, now that the remaining PRs have been merged first.

I know that the documentation still needs some work, will work on that.

https://github.com/emgre/win-crypto-ng/pull/27/commits/bcf5808bfbb2aaa2c70e437935fda44a682add8a (Handle more CNG errors)  - this is unrelated and is merely a nice addition, so can be merged independently

https://github.com/emgre/win-crypto-ng/pull/27/commits/bcf5808bfbb2aaa2c70e437935fda44a682add8a (WIP: Implement importing keys just from blob payload itself) - I think that's the biggest issue with this PR. To use a low-level key import facility, we need to create a raw `Blob` from parts first (which may in turn require some `winapi` types). That commit introduces explicit `import_from_parts` function impls for relevant key type. Not pretty, but it does the job.

Another possibility is to tie the `import_from_parts` function argument somehow to the 'payload' part of the key blob via traits but I'm not sure if it's ergonomic (e.g. ECDH Curve25519 key only needs to accept a compressed Y coord but the generic ECC key blob specifies both X and Y) nor possible (not sure if all blob headers can be derived from the payload itself, e.g. DSA v1/v2 format with custom generator values and whatnot)

https://github.com/emgre/win-crypto-ng/pull/27/commits/c5dd3b30bc4fb370211160a5c91c407ecaf8db56 (Implement simple secret agreement functionality) - this doesn't implement all key derivation functions/modes but rather only the raw secret agreement value. We can implement the rest in an another PR.

Xanewok committed:
Polish public interface and add missing documentation

Xanewok committed:
Mention new functionality to README.md

Xanewok commented:
Rebased on top of https://github.com/emgre/win-crypto-ng/commit/12b5824fe5329eeff47f3d70345d30ff6b411378, added quite a bit of documentation (for asymmetric stuff) and did some final tweaking (e.g. revamped some visibility stuff like accidentally hidden struct members, sealed a trait).

I didn't touch the `import_from_parts` and I currently don't have an idea how to make that more ergonomic and possibly unified with the `Import` trait. If you don't mind, I'd love to ship this in 0.2 as-is and think about how we can make that nicer in the subsequent 0.* releases.

I can squash the current follow-up commits in the end, I just wanted to make it easier to see what's been added since the last change to make it easier to review.

emgre commented on src/asymmetric/agreement.rs:
```diff
@@ -0,0 +1,115 @@
+//! Functionality related to secret agreement and key derivation.
+//!
+//! > **NOTE**: This is currently a stub and should be expanded.
```
This comment is obsolete.

Xanewok replied:
Removed

emgre commented on src/asymmetric/builder.rs:
```diff
@@ -0,0 +1,724 @@
+//! use win_crypto_ng::asymmetric::{AsymmetricAlgorithm, AsymmetricAlgorithmId, AsymmetricKey};
+//! use win_crypto_ng::asymmetric::{ecc::NistP384, Ecdh, Rsa};
+//!
+//! // Handle to an asymmetric key whose algorithm ID is known at run-time
```
I would add a textual explanation of the dynamic algorithm vs. the typed algorithm because it's not the most obvious thing to understand.

Xanewok replied:
Added a new 'Type-safety' section to `AsymmetricKey` that explains the reason/adds an example

emgre reviewed (dismissed):
Made some minor comments in the code.

I still think we should separate the huge `asymmetric/mod.rs` into each algorithm to make it clearer what is going on.

As you mentioned, the import/export would need some more thoughts.

The asymmetric module should have more extensive unit tests, but I guess Sequoia will be the guinea pig for it.

Since you need this, I don't mind merging it as is and provide a release on crates.io. Be advised that the release after that will certainly include breaking changes.

Xanewok commented:
Squashed the final commits and addressed the review comments.

Changes between those two revisions:
`$ git diff c9705fe 7a65508f`
```diff
diff --git a/src/asymmetric/agreement.rs b/src/asymmetric/agreement.rs
index dedf3a6..5b00431 100644
--- a/src/asymmetric/agreement.rs
+++ b/src/asymmetric/agreement.rs
@@ -1,6 +1,4 @@
 //! Functionality related to secret agreement and key derivation.
-//!
-//! > **NOTE**: This is currently a stub and should be expanded.
 
 use crate::asymmetric::{ecc::Curve, AsymmetricKey, Ecdh, Private, Public};
 use crate::helpers::Handle;
diff --git a/src/asymmetric/mod.rs b/src/asymmetric/mod.rs
index 7800271..b61ac8e 100644
--- a/src/asymmetric/mod.rs
+++ b/src/asymmetric/mod.rs
@@ -245,17 +245,38 @@ impl Algorithm for AsymmetricAlgorithmId {
 /// Can be specified as that of which type is known at run-time (erased) or
 /// with a concrete algorithm parameter.
 ///
+/// ## Type-safety
+/// At the time of writing, all of the relevant crypto operations are
+/// implemented for only keys whose algorithm is statically known. This means
+/// that the keys using the dynamic `AsymmetricAlgorithmId` need to be
+/// re-imported in order to statically verify the algorithm used.
+///
+/// ```
+/// use win_crypto_ng::asymmetric::{AsymmetricAlgorithm, AsymmetricAlgorithmId, AsymmetricKey};
+/// use win_crypto_ng::asymmetric::{Rsa, Private, Import, Export};
+///
+/// let dynamic: AsymmetricKey<AsymmetricAlgorithmId, Private> =
+///     AsymmetricKey::builder(AsymmetricAlgorithmId::Rsa).key_bits(512).build().unwrap();
+/// let dynamic_blob = dynamic.export().unwrap();
+///
+/// let provider = AsymmetricAlgorithm::open(AsymmetricAlgorithmId::Rsa).unwrap();
+/// let rsa_blob = dynamic_blob.try_into().unwrap_or_else(|_| panic!());
+/// let rsa_key = AsymmetricKey::<Rsa, Private>::import(Rsa, &provider, &rsa_blob).unwrap();
+/// // Can now additional impls for `AsymmetricKey<Rsa, Private>`, e.g.
+/// // rsa_key.decrypt(..);
+/// ```
+///
 /// ## Generation
 /// ```
 /// use win_crypto_ng::asymmetric::{AsymmetricAlgorithm, AsymmetricAlgorithmId, AsymmetricKey};
-/// use win_crypto_ng::asymmetric::Rsa;
+/// use win_crypto_ng::asymmetric::{Rsa, Private};
 ///
 /// // Handle to an asymmetric key whose algorithm ID is known at run-time
-/// let handle: AsymmetricKey<AsymmetricAlgorithmId, _> =
+/// let handle: AsymmetricKey<AsymmetricAlgorithmId, Private> =
 ///     AsymmetricKey::builder(AsymmetricAlgorithmId::Rsa).key_bits(512).build().unwrap();
 ///
 /// // Handle to an asymmetric key whose algorithm is known to be RSA
-/// let handle: AsymmetricKey<Rsa, _> =
+/// let handle: AsymmetricKey<Rsa, Private> =
 ///     AsymmetricKey::builder(Rsa).key_bits(512).build().unwrap();
 /// ```
 pub struct AsymmetricKey<A: Algorithm = AsymmetricAlgorithmId, P: Parts = Public>(
@@ -929,7 +950,9 @@ impl<P: Parts> AsymmetricKey<Rsa, P> {
             .map(|_| output.into_boxed_slice())
         }
     }
+}
 
+impl AsymmetricKey<Rsa, Private> {
     /// Decrypt a message using the RSA key.
     pub fn decrypt(&self, padding: Option<EncryptionPadding>, data: &[u8]) -> Result<Box<[u8]>> {
         let mut out = WindowsString::new();
```

The final change is implementing `decrypt` only on the key having `Private` parts (but we can still encrypt using private/public keys).

I agree about splitting the algorithm section but as I said, I'd like to do that in the next release.

@emgre could you please merge this PR and publish a 0.2 version (symmetric encryption padding introduced breaking changes) containing these changes? Thank you!
